File: src/main/java/ticklock/entity/EventEntity.java
package ticklock.entity;

import jakarta.persistence.*;

@Entity
@Table(name = "events")
public class EventEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private String name;

    private int totalSeats;

    private int remainingSeats;

    @Version
    private Long version;

    protected EventEntity() {
    }

    public EventEntity(String name, int totalSeats) {
        this.name = name;
        this.totalSeats = totalSeats;
        this.remainingSeats = totalSeats;
    }

    public Long getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public int getTotalSeats() {
        return totalSeats;
    }

    public int getRemainingSeats() {
        return remainingSeats;
    }

    public Long getVersion() {
        return version;
    }

    public boolean hasRemainingSeats() {
        return remainingSeats > 0;
    }

    public void decreaseSeat() {
        if (remainingSeats <= 0) {
            throw new IllegalStateException("남은 좌석이 없습니다.");
        }
        this.remainingSeats--;
    }
}

File: src/main/kotlin/ticklock/entity/EventEntity.kt
package ticklock.entity

import jakarta.persistence.*

@Entity
@Table(name = "events")
class EventEntity(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,

    val name: String,

    val totalSeats: Int,

    var remainingSeats: Int = totalSeats,

    @Version
    val version: Long? = null
) {
    constructor(name: String, totalSeats: Int) : this(
        id = null,
        name = name,
        totalSeats = totalSeats,
        remainingSeats = totalSeats,
        version = null
    )
    fun hasRemainingSeats(): Boolean = remainingSeats > 0

    fun decreaseSeat() {
        if (remainingSeats <= 0) {
            throw IllegalStateException("남은 좌석이 없습니다.")
        }
        remainingSeats--
    }
}